homeassistant/components/tahoma/sensor.py

```diff
@@ -0,0 +1,153 @@
+CORE_FOSSIL_ENERGY_CONSUMPTION_STATE = "core:FossilEnergyConsumptionState"
+CORE_GAS_CONSUMPTION_STATE = "core:GasConsumptionState"
+CORE_LUMINANCE_STATE = "core:LuminanceState"
+CORE_MEASURED_VALUE_TYPE = "core:MeasuredValueType"
+CORE_RELATIVE_HUMIDITY_STATE = "core:RelativeHumidityState"
+CORE_SUN_ENERGY_STATE = "core:SunEnergyState"
+CORE_TEMPERATURE_STATE = "core:TemperatureState"
+CORE_THERMAL_ENERGY_CONSUMPTION_STATE = "core:ThermalEnergyConsumptionState"
+CORE_WATER_CONSUMPTION_STATE = "core:WaterConsumptionState"
+CORE_WINDSPEED_STATE = "core:WindSpeedState"
+
+UNIT_LX = "lx"
+
+TAHOMA_SENSOR_DEVICE_CLASSES = {
+    "CO2Sensor": None,  # co2
+    "COSensor": None,  # co
+    "ElectricitySensor": DEVICE_CLASS_POWER,
+    "HumiditySensor": DEVICE_CLASS_HUMIDITY,
+    "LightSensor": DEVICE_CLASS_ILLUMINANCE,
+    "RelativeHumiditySensor": DEVICE_CLASS_HUMIDITY,
+    "SunSensor": None,  # sun_energy
+    "TemperatureSensor": DEVICE_CLASS_TEMPERATURE,
+    "WindSensor": None,  # wind_speed
+}
+
+# From https://www.tahomalink.com/enduser-mobile-web/steer-html5-client/tahoma/bootstrap.js
+UNITS = {
+    "core:TemperatureInCelcius": TEMP_CELSIUS,
+    "core:TemperatureInCelsius": TEMP_CELSIUS,
+    "core:TemperatureInKelvin": TEMP_KELVIN,
+    "core:TemperatureInFahrenheit": TEMP_FAHRENHEIT,
+    "core:LuminanceInLux": UNIT_LX,
+    "core:ElectricCurrentInAmpere": ELECTRICAL_CURRENT_AMPERE,
+    "core:VoltageInVolt": VOLT,
+    "core:ElectricalEnergyInWh": ENERGY_WATT_HOUR,
+    "core:ElectricalEnergyInKWh": ENERGY_KILO_WATT_HOUR,
+    "core:ElectricalEnergyInMWh": f"M{ENERGY_WATT_HOUR}",
+    "core:ElectricalPowerInW": POWER_WATT,
+    "core:ElectricalPowerInKW": POWER_KILO_WATT,
+    "core:ElectricalPowerInMW": f"M{POWER_WATT}",
+    "core:FlowInMeterCubePerHour": VOLUME_CUBIC_METERS,
+    "core:LinearSpeedInMeterPerSecond": SPEED_METERS_PER_SECOND,
+    "core:RelativeValueInPercentage": PERCENTAGE,
+    "core:VolumeInCubicMeter": VOLUME_CUBIC_METERS,
+    "core:VolumeInLiter": VOLUME_LITERS,
+    "core:FossilEnergyInWh": ENERGY_WATT_HOUR,
+    "core:FossilEnergyInKWh": ENERGY_KILO_WATT_HOUR,
+    "core:FossilEnergyInMWh": f"M{ENERGY_WATT_HOUR}",
+    "meters_seconds": SPEED_METERS_PER_SECOND,
+}
+
+UNITS_BY_DEVICE = {
+    "CO2Sensor": CONCENTRATION_PARTS_PER_MILLION,
+    "COSensor": CONCENTRATION_PARTS_PER_MILLION,
+}
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up the TaHoma sensors from a config entry."""
+    data = hass.data[DOMAIN][entry.entry_id]
+    coordinator = data["coordinator"]
+
+    entities = [
+        TahomaSensor(device.deviceurl, coordinator)
+        for device in data["platforms"][SENSOR]
+        if device.states
+    ]
+
+    async_add_entities(entities)
+
+
+class TahomaSensor(TahomaEntity, Entity):
+    """Representation of a TaHoma Sensor."""
+
+    @property
+    def state(self):
+        """Return the value of the sensor."""
+        state = self.executor.select_state(
```

Comment:
Can we avoid having to search for which kind it is every time we write state to the state machine?

Reply:
Would you prefer to loop through the list once on init and set the available state name in the class? We thought it wouldn't hurt too much since it is just a very small dictionary with states.

Reply:
That sounds better. Probably do it in `async_added_to_hass`

Reply:
We will have a look here! This will be applicable for most of the platforms (e.g. `binary_sensor`, `cover` and `switch`). To be honest, I don't think that the performance hit will be visible, I could have a look with `timeit` as well to see the impact.

Our initial approach was to have it as dynamic as possible, since the current tahoma integration has many things hardcoded which limits device support. For more complex devices like climate and water_heater we have split them to different entities.

Reply:
It looks like everything is indexed by `name`. Is it possible to have two states with the same name?  Does a `dict` make more sense to store the states instead of having to do linear searches each time you need to find the state?

Reply:
@bdraco could make sense, currently we just use the original API output. This would mean that we need to rewrite the states to a dict instead of a list on every API request.

@tetienne, @vlebourl what do you think? Or any other suggestions based on your experiences with the API?

Reply:
If I understand correctly, the suggestion is to create a dict on two levels: command -> widgeui -> state name. It’s not something we can do easily. 
But, what about caching the result of [select_state](https://github.com/iMicknl/home-assistant-core/blob/feature/tahoma/add_sensor_platform/homeassistant/components/tahoma/overkiz_executor.py#L34)?
We can give the command name to the function (state, current_cover_position, set_position, etc). 

```python
def select_state(self, command_name, *states) -> Optional[str]:
        """Select first existing active state in a list of states."""
        if command not in self.command_state and  self.device.states:
            state = next(
                (
                    state.value
                    for state in self.device.states
                    if state.name in list(states)
                ),
                None,
            )
            self.command_state[command] = state
        return self.command_state.get(command)
```

Reply:
https://github.com/iMicknl/python-tahoma-api/blob/1846fa03388ed56cc470cf08952889d17bcaa0c1/pyhoma/models.py#L158

Can the underlying class store it in a `dict` indexed by `state.name`?

It looks like you have support for `__getitem__` so could use that here instead and optimize the underlying library later

Reply:
Thanks @bdraco, I will look into this in the upcoming days. I agree that we really need to fix this for all platforms in the integration.

My only doubt is still that we tried to write our code as dynamically as possible for `sensor`, `binary_sensor` and `cover` since we have to support over 6000 devices... I guess a good way forward would be to cache the first available state / command, but not sure where yet.